test/tokenizedSubaccounts/PPTSA/PPTSA_Validation.t.sol

```diff
@@ -0,0 +1,502 @@
+
+    action = _createWithdrawalAction(3e18);
+    vm.prank(signer);
+    vm.expectRevert(PrincipalProtectedTSA.PPT_WithdrawingWithOpenTrades.selector);
+    tsa.signActionData(action, "");
+
+    vm.warp(block.timestamp + 8 days);
+    _setSettlementPrice("weth", uint64(expiry), 1500e18);
+    srm.settleOptions(markets["weth"].option, tsa.subAccount());
+
+    vm.startPrank(signer);
+    // now try to withdraw all of the base asset. Should fail
+    action = _createWithdrawalAction(3e18);
+    vm.expectRevert(PrincipalProtectedTSA.PPT_WithdrawingUtilisedCollateral.selector);
+    tsa.signActionData(action, "");
+
+    // now try to a small 5% of the base asset. Should pass
+    action = _createWithdrawalAction(0.15e18);
+    tsa.signActionData(action, "");
+
+    vm.stopPrank();
+  }
+
+  /////////////////
+  //    Trades   //
+  /////////////////
+  function testTradeValidation() public {
+    _depositToTSA(10e18);
+    _executeDeposit(10e18);
+
+    // Receive positive cash from selling options
+    uint64 expiry = uint64(block.timestamp + 7 days);
+    _tradeRfqAsMaker(1e18, 1e18, expiry, 400e18, 4e18, 800e18, true);
+
+    (uint openSpreads, uint base, int cash) = tsa.getSubAccountStats();
+    assertEq(openSpreads, 1e18);
+    assertEq(base, 10e18);
+    assertEq(cash, 3e18);
+
+    ITradeModule.TradeData memory tradeData = ITradeModule.TradeData({
+      asset: address(markets["weth"].base),
+      subId: OptionEncoding.toSubId(expiry, 2200e18, true),
+      limitPrice: int(1e18),
+      desiredAmount: 2e18,
+      worstFee: 1e18,
+      recipientId: tsaSubacc,
+      isBid: true
+    });
+
+    IActionVerifier.Action memory action = IActionVerifier.Action({
+      subaccountId: tsaSubacc,
+      nonce: ++tsaNonce,
+      module: tradeModule,
+      data: abi.encode(tradeData),
+      expiry: block.timestamp + 8 minutes,
+      owner: address(tsa),
+      signer: address(tsa)
+    });
+
+    vm.startPrank(signer);
+
+    tradeData.desiredAmount = 0;
+    action.data = abi.encode(tradeData);
+    vm.expectRevert(PrincipalProtectedTSA.PPT_InvalidDesiredAmount.selector);
+    tsa.signActionData(action, "");
+
+    tradeData.desiredAmount = 2.0e18;
+    action.module = IMatchingModule(address(10));
+    action.data = abi.encode(tradeData);
+    vm.expectRevert(PrincipalProtectedTSA.PPT_InvalidModule.selector);
+    tsa.signActionData(action, "");
+
+    action.module = tradeModule;
+    tradeData.asset = address(markets["weth"].option);
+    action.data = abi.encode(tradeData);
+    vm.expectRevert(PrincipalProtectedTSA.PPT_InvalidAsset.selector);
+    tsa.signActionData(action, "");
+
+    tradeData.asset = address(markets["weth"].base);
+    action.data = abi.encode(tradeData);
+    tsa.signActionData(action, "");
+
+    vm.warp(block.timestamp + 1 days);
+    vm.expectRevert(PrincipalProtectedTSA.PPT_InvalidActionExpiry.selector);
+    tsa.signActionData(action, "");
+  }
+
+  /////////////////
+  //     RFQ     //
+  /////////////////
```

Comment:
need tests somewhere that:
- ensure the `executor` doesn't open too many spreads
- ensure `executor` doesn't open spreads when there is `cashBalance is negative` && `cashBalance.abs() < negMaxCash`
- might have missed it, but don't see all 8 tests on the full metrix of potential vault params e.g. `isLongSpread` * `isCall` * `isMaker`

Reply:
Talked about this offline. A good amount of the tests have changed now so should resolve the comments.